lib/HTFeed/DataDenZipAudit.pm

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/perl
+package HTFeed::DataDenZipAudit;
+
+use strict;
+use warnings;
+use HTFeed::DBTools qw(get_dbh);
+use HTFeed::Volume;
+use Carp;
+use Log::Log4perl;
+
+# Procedural function for randomly choosing a volume to check.
+#
+# my $vol = HTFeed::DataDenZipAudit->choose();
+# my $audit = HTFeed::DataDenZipAudit->new(namespace => $vol->{namespace},
+#                                          objid => $vol->{objid},
+#                                          path => $vol->{path},
+#                                          version => $vol->{version});
+# my $result = $audit->run();
+sub choose {
+  my $path_prefix = shift || '/htdataden';
+
+  my $sql = 'SELECT COUNT(*) FROM feed_backups' .
+            " WHERE path LIKE '$path_prefix%'" .
+            ' AND deleted IS NULL AND saved_md5sum IS NOT NULL';
+  my $dbh = HTFeed::DBTools->get_dbh();
+  my ($rows) = $dbh->selectrow_array($sql);
+  my $random_row = int(rand($rows));
+  my ($namespace, $objid, $path, $version);
+  $sql = 'SELECT namespace,id,path,version FROM feed_backups' .
+         " WHERE path LIKE '$path_prefix%'" .
+         ' AND deleted IS NULL' .
+         ' AND saved_md5sum IS NOT NULL' .
+         ' LIMIT ?,1';
+  eval {
+    ($namespace, $objid, $path, $version) = $dbh->selectrow_array($sql, undef, $random_row);
+  };
+  if ($@) {
+    die "Database query failed: $@";
+  }
+  return { namespace => $namespace, objid => $objid,
+           path => $path, version => $version };
+}
+
+sub new {
+  my $class = shift;
+
+  my $object = {
+      @_,
+  };
+  if ( $class ne __PACKAGE__ ) {
+      croak "use __PACKAGE__ constructor to create $class object";
+  }
+  # check parameters
+  unless ($object->{namespace} && $object->{objid} &&
+          $object->{version} && $object->{path}) {
+    croak "invalid args: namespace, objid, version, path required";
+  }
+  $object->{mets_path} = $object->{path} . '/' . $object->{objid} . '.mets.xml';
+  $object->{zip_path} = $object->{path} . '/' . $object->{objid} . '.zip.gpg';
+  
+  my $volume = HTFeed::Volume->new(
+    packagetype => 'pkgtype',
+    namespace   => $object->{namespace},
+    objid       => $object->{objid}
+  );
+  $object->{volume} = $volume;
+  # EVIL HACK that may break in production.
+  $object->{storage} ||= (HTFeed::Storage::for_volume($volume))[0];
```

Comment:
Yeah, I think what we need here is a name for the particular instance of the storage back end, both in the config and in the database, so we can instantiate the correct one. There is a ticket (https://tools.lib.umich.edu/jira/browse/HT-2812) for that -- perhaps we should queue that up for next sprint?

Reply:
Works for me. I'm gonna leave this part of the code alone for now since it does work with its rickety test scaffolding.